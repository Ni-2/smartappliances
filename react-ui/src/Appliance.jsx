import React from 'react';
import axios from 'axios';
import _ from 'lodash';
import Task from './Task';
import Description from './Description';

const shapeInfo = (info) =>
  Object.entries(info).map(([title, chapter]) => (
    <div key={_.uniqueId()}>
      <h4>{title}</h4>
      <ul>
        {Array.isArray(chapter)
          ? chapter.map((note) => <li key={_.uniqueId()}>{note}</li>)
          : Object.entries(chapter).map(([feature, data]) => (
              <li key={_.uniqueId()}>
                <b>{feature}</b> {data}
              </li>
            ))}
      </ul>
    </div>
  ));

export default function Appliance(props) {
  const { url, id, setCards, toggleToAppliance, value } = props;
  const { image, name, usersDescription, applState, tasks, info } = value;

  const handleDeleteAppl = async (e) => {
    e.preventDefault();
    const response = await axios.delete(url, {
      headers: { 'Content-Type': 'application/json' },
      params: { serial: id },
    });
    toggleToAppliance(null);
    setCards(response.data);
  };

  return (
    <div className="container appliance">
      <div className="row">
        <div className="col-12 col-md-6">
          <div className="appliance-img-container m-t-2rem">
            <img src={image} className="appliance-img" alt={name} />
          </div>
          <Task url={url} id={id} setCards={setCards} applState={applState} tasks={tasks} />
        </div>
        <div className="col-12 col-md-6 m-t-1rem">
          <h2 className="title">{name}</h2>
          <Description url={url} id={id} setCards={setCards} usersDescription={usersDescription} />
          <div className="m-t-1rem">{shapeInfo(info)}</div>
          <button type="button" className="btn btn-danger" onClick={handleDeleteAppl}>
            Delete appliance
          </button>
        </div>
      </div>
    </div>
  );
}
