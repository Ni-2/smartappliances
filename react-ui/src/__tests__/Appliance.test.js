import React from 'react';
import renderer from 'react-test-renderer';
import { promises as fsp } from 'fs';
import path from 'path';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import { render, fireEvent, waitFor, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import Appliance from '../Appliance';

it('renders form', () => {
  const component = renderer.create(
    <Appliance
      value={{
        image: 'image',
        name: 'name',
        usersDescription: 'usersDescription',
        applState: 'applState',
        tasks: 'tasks',
        info: 'info',
      }}
      toggleToAppliance={() => 'function toggleToAppliance'}
      url="/api"
      id="101"
      setCards={() => 'function setCards'}
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

describe('interactions', () => {
  const getFixturePath = (filename) => path.resolve(__dirname, '__fixtures__', filename);
  const readFixtureFile = (filename) => fsp.readFile(getFixturePath(filename), 'utf-8');

  let cards;
  const server = setupServer(
    rest.post('/api/newTask', (req, res, ctx) => {
      const { id } = req.body;
      return id === '101' && res(ctx.status(201), ctx.json(cards));
    }),
  );

  beforeAll(async () => {
    const cardsStringified = await readFixtureFile('cards-101-201-301.json');
    cards = JSON.parse(cardsStringified);
    server.listen();
  });

  afterEach(() => server.resetHandlers());

  afterAll(() => server.close());

  it('sends task to the server', async () => {
    render(
      <Appliance
        value={cards['101']}
        toggleToAppliance={() => 'function toggleToAppliance'}
        url="/api"
        id="101"
        setCards={() => cards}
      />,
    );

    expect(screen.getByRole('combobox')).toHaveValue('default');
    fireEvent.change(screen.getByRole('combobox'), { target: { value: 'wool' } });
    expect(screen.getByRole('combobox')).toHaveValue('wool');

    fireEvent.click(screen.getByText('Start'));
    await waitFor(() => expect(screen.getByRole('combobox')).toHaveValue('default'));
  });
});
