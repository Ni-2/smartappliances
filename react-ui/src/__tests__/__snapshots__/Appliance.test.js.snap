// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders form 1`] = `
<div
  className="container appliance"
>
  <div
    className="row"
  >
    <div
      className="col-12 col-md-6"
    >
      <div
        className="appliance-img-container m-t-2rem"
      >
        <img
          alt="name"
          className="appliance-img"
          src="image"
        />
      </div>
      <div
        className="flex m-t-2rem"
      >
        <form
          className="row"
          onSubmit={[Function]}
        >
          <div
            className="col-12 m-1rem"
          >
            <h4
              className="current-state"
            >
              Current state: 
              <span
                className="green"
              >
                applState
              </span>
            </h4>
            <label
              className="form-label"
              htmlFor="inputState"
            >
              <h4>
                Add task
              </h4>
            </label>
            <select
              className="form-select"
              id="inputState"
              onChange={[Function]}
              value="default"
            >
              <option
                value="default"
              >
                Choose...
              </option>
              <option
                value="0"
              >
                t
              </option>
              <option
                value="1"
              >
                a
              </option>
              <option
                value="2"
              >
                s
              </option>
              <option
                value="3"
              >
                k
              </option>
              <option
                value="4"
              >
                s
              </option>
            </select>
          </div>
          <div
            className="col-12 m-1rem"
          >
            <button
              className="btn btn-primary"
              type="submit"
            >
              Start
            </button>
          </div>
        </form>
      </div>
    </div>
    <div
      className="col-12 col-md-6 m-t-1rem"
    >
      <h2
        className="title"
      >
        name
      </h2>
      <div
        className="flex al-i-base f-s-1-2rem"
      >
        <p
          className="text m-1rem"
        >
          <b>
            description: 
          </b>
          usersDescription
        </p>
        <button
          aria-controls="changeDescription"
          aria-expanded="false"
          className="btn btn-outline-secondary m-1rem f-s-1-2rem"
          data-bs-target="#changeDescription"
          data-bs-toggle="collapse"
          type="button"
        >
          Change my description
        </button>
      </div>
      <div
        className="collapse m-lr-1rem"
        id="changeDescription"
      >
        <div
          className="flex"
        >
          <form
            className="row"
            onSubmit={[Function]}
          >
            <div
              className="col-auto"
            >
              <label
                className="visually-hidden"
                htmlFor="inputDescription"
              >
                New description
              </label>
              <input
                className="form-control"
                id="inputDescription"
                onChange={[Function]}
                placeholder="new description"
              />
            </div>
            <div
              className="col-auto"
            >
              <button
                className="btn btn-primary mb-3"
                type="submit"
              >
                Without description
              </button>
            </div>
          </form>
        </div>
      </div>
      <div
        className="m-t-1rem"
      >
        <div>
          <h4>
            0
          </h4>
          <ul>
            <li>
              <b>
                0
              </b>
               
              i
            </li>
          </ul>
        </div>
        <div>
          <h4>
            1
          </h4>
          <ul>
            <li>
              <b>
                0
              </b>
               
              n
            </li>
          </ul>
        </div>
        <div>
          <h4>
            2
          </h4>
          <ul>
            <li>
              <b>
                0
              </b>
               
              f
            </li>
          </ul>
        </div>
        <div>
          <h4>
            3
          </h4>
          <ul>
            <li>
              <b>
                0
              </b>
               
              o
            </li>
          </ul>
        </div>
      </div>
      <button
        className="btn btn-danger"
        onClick={[Function]}
        type="button"
      >
        Delete appliance
      </button>
    </div>
  </div>
</div>
`;
